### PR TITLE
Throw exceptions when attempting to ingest an already existing datastream.

### PR DESCRIPTION
**Pull request by jordandukart**
**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1995

# What does this Pull Request do?
Changes unexpected behavior in Tuque where ingesting a datastream on an object that already has that datastream causes a `FALSE` to be returned rather than an exception thrown. This gets propagated up to the Tuque wrapper in Islandora and causes `islandora_datastream_ingested` hooks to be invoked on a non-existent `NewFedoraDatastream`.

# What's new?
Exception being tossed in the event a datastream is attempted to be added to a `NewFedoraObject` or `FedoraObject` that already exists on the object. Similarly, doing this uncovered a bug in the `FedoraRelationships` class that could lead to odd caching behavior when building up relationships on a `NewFedoraObject`.

# How should this be tested?
Enable the test module from: https://github.com/jordandukart/datastream_ingest_test
Re-ingest a collection that defines a TN from the required objects solution packs page (admin/islandora/solution_pack_config/solution_packs)
Note how the messages show that the Tuque wrapper from Islandora is invoking hooks with a `NewFedoraDatastream`
Pull Tuque code.
Repeat the process, note how we are now tossing an exception.

# Additional Notes:
Relates to https://github.com/Islandora/islandora/pull/678 and https://github.com/Islandora/islandora_solution_pack_collection/pull/191

* Could this change impact execution of existing code? Yes.
It's possible that somewhere someone is relying upon these ingestDatastream calls returning `FALSE` as opposed to exceptions.

# Interested parties
@DiegoPino, @dannylamb, @jonathangreen

### CODE REVIEW

File: Object.php
Comment (willtp87) [inline Object.php]: Can we fix this while we are here? should be checking DS not object.

Comment (willtp87) [reply Object.php]: Perhaps it should just be checking if the logMessage is set or not.

Comment (willtp87) [reply Object.php]: Maybe both.

**Review by willtp87 — APPROVED** [review]

**Comment by jordandukart** [comment]
Bump @Islandora/7-x-1-x-committers

**Comment by jordandukart** [comment]
Bump @Islandora/7-x-1-x-committers

**Comment by DiegoPino** [comment]
@jordandukart i really need to test this and a bit short of time. I know i know it must be super urgent, promise i will deal with it today. Thanks

**Comment by jordandukart** [comment]
Not overly urgent @DiegoPino just didn't see a comment as such!

**Comment by jordandukart** [comment]
Just FYI @DiegoPino there needs to be some work done on this for a deprecation cycle as discussed on committers so I'll ping you again once that's done.

**Comment by jordandukart** [comment]
Updated https://github.com/Islandora/islandora/pull/680 to have a deprecation cycle as discussed on committers. Spun https://github.com/Islandora/islandora_paged_content/pull/130 and https://github.com/Islandora/islandora_ocr/pull/80/files from that as well.

**Comment by DiegoPino** [comment]
@jordandukart ok, i got this. Really need to test this thing before merging. Will have it done (and all its befriended pulls) by the end of the day

File: Object.php
Comment (adam-vessey) [inline Object.php]: Comment needs to be updated?

**Comment by DiegoPino** [comment]
@jordandukart @adam-vessey do you have still internal review pending? If so ping me, don't want to get in your way || double review.

**Comment by DiegoPino** [comment]
@jordandukart i restarted travis because it was failing . Will merge once you confirm there is not more work pending on the other pulls because i feel this all needs to go together.

**Review by adam-vessey — CHANGES_REQUESTED** [review]
Looks good here. (I clicked the wrong thing... "Request changes" instead of "Approve"... whoops.)

**Review by adam-vessey — APPROVED** [review]

**Comment by jordandukart** [comment]
Bump @Islandora/7-x-1-x-committers

**Comment by jordandukart** [comment]
Bump @Islandora/7-x-1-x-committers, @DiegoPino

**Comment by DiegoPino** [comment]
@jordandukart can you believe i only saw the "requested changes" email notification from @adam-vessey and forgot about this. Now i'm seeing it was just a wrong button thing! :man_facepalming: Will look into it now.

**Comment by DiegoPino** [comment]
@jordandukart will merge. Seems to be ok. Will start with this one and then move to https://github.com/Islandora/islandora/pull/678 ~~and finish with https://github.com/Islandora/islandora_solution_pack_collection/pull/191~~ i merged that one already

By the way, what about this one here? Not sure if directly directly related, but related to your other pulls..

https://github.com/Islandora/islandora_solution_pack_compound/blob/3645b627647410e6ddc9326e67886c4638f7768b/islandora_compound_object.module#L340

Are we there checking for (empty($parent['TN'])) and then constructing and ingesting the datastream? empty and isset are not the same ... mmm... 

Could you guys make sure we have an isset in front on all ingest datastream operations on core? Or not needed in your opinion?

**Comment by jordandukart** [comment]
I can have a look but seem equivalence for empty should be okay http://php.net/manual/en/function.empty.php?

**Comment by DiegoPino** [comment]
@jordandukart cool. Not sure really, but i can be wrong.  I have seen failed ingested data streams to eval to empty (because they have "null" content) but the data stream entry actually exists, which differs from a nonexisting dsid key in the object´s datastream accessor. Maybe  just paranoid here.

**Pull request merged**
